performjoin.go

```diff
@@ -0,0 +1,276 @@
+package gomatrixserverlib
+
+import (
+	"context"
+	"crypto/ed25519"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/gomatrixserverlib/spec"
+	"github.com/sirupsen/logrus"
+)
+
+type PerformJoinInput struct {
+	UserID     *spec.UserID
+	RoomID     string
+	ServerName spec.ServerName
+	Content    map[string]interface{}
+	Unsigned   map[string]interface{}
+
+	PrivateKey ed25519.PrivateKey
+	KeyID      KeyID
+	KeyRing    *KeyRing
+
+	EventProvider EventProvider
+}
+
+type PerformJoinResponse struct {
+	JoinEvent     *HeaderedEvent
```

Comment:
Please use `*Event` as it'll make merge conflicts easier to wrangle for me. You have the same data on `*Event`, so use `JoinEvent.Version()` to get the `RoomVersion`.